### ISSUE TITLE
Unicode string length of code points vs grapheme length

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#1402]
From this blog post about Strings in Swift 2:

https://developer.apple.com/swift/blog/?id=30

We get some test cases that are weird:

``` factor
IN: scratchpad "cafe" "\u000301" append dup length

--- Data stack:
"café"
5
```

We don't do any unicode normalization when we treat strings as character sequences and append unicode accents...

**Comment by mrjbq7** [comment]
I think this is also the problem with #1382.

**Comment by mrjbq7** [comment]
Stuff like this with emoji:

```factor
IN: scratchpad "🇪🇸" length .
2
```

**Comment by pirj** [comment]
It's however impossible to normalize 🇪🇸, it is a sequence of two unicode code points that should be displayed as a single glyph. It depends on the font used if it is displayed as one or two symbols.
I see this as two distinct problems:
 - we need a way to know how long the string is in glyphs to be able to properly align cursor (#1382)
 - we need to make sure sequence words for string type use canonical equivalence (this issue)

PS By the way I wouldn't say that flag is properly aligned to surrounding symbols when writing this comment (OSX/Chrome).
PPS JS determines the length of "🇪🇸" as 4

**Comment by erg** [comment]
One problem is we're on Unicode 5.2 when we could be on Unicode 9.0...

Also, the data files are interspersed in the source tree for some reason. They should just go into one Unicode Data directory.

**Comment by bjourne** [comment]
Unicode is crayze. Let's just redefine string length to be the count of the number of code points. :fish_cake:

**Comment by kusumotonorio** [comment]
IN: scratchpad USE: unicode "cafe" "\u000301" append dup >graphemes length

--- Data stack:
"café"
4

It seems to me that Factor already has the ability to handle graphemes, so  the problem is to comply with the new unicode standard.

**Comment by erg** [comment]
I have a branch called ``modern-harvey3`` with a bit of unicode work. The goal of the branch is to make the parser have simple rules instead of parsing words, but that part is not finished.

I updated the unicode files just now. Looks like with unicode 12.10.1 there are many failing collation tests but it should be pretty easy to work on and fix them.

``"unicode" test``

It probably isn't that hard to fix them all and then backport the unicode directory to master. The only major change might be changing back ``ch'a`` to ``CHAR: a``, which was an experiment that I'm not big on changing anymore.

**Comment by kusumotonorio** [comment]
I'm glad to hear some great information about the Unicode update!
I'm looking forward to the new parser.

**Comment by kusumotonorio** [comment]
I am trying to backport your unicode improvement.
What I did was do the following in the unicode directory:
- change from "ch'a" to "CHAR: a"
- change from " \CATEGORY:" or " \CATEGORY-NOT:" to "CATEGORY:" or "CATEGORY - NOT:"
- change from "postpone:" to "POSTPONE:"
- comment out some documents in error
- changes to a few error point

I replaced the unicode directory, but when I inputted emojis in a listener window and moved the caret, it didn't work as expected.
If I can backport correctly, will I be able to handle emojis properly in the listener window?

**Comment by mrjbq7** [comment]
To fix the listener issues, we need to change it to step through a string glyph by glyph and not code point by code point. 

I don’t think @erg did anything with that yet. Perhaps we need a more Unicode aware string type, maybe Go or Python can provide some inspiration.

**Comment by kusumotonorio** [comment]
For example, 😀 (U+1F600) is an emoji expressed with a single code point. In that respect, it is the same as a regular alphabet. However, the current Factor does not handle it well on the editor.

On the other hand, Factor properly handles the "é" represented by two code points on the editor.

This means that Factor already has a mechanism to handle each glyph (grapheme cluster). And it seems that characters that were not expected when the current unicode support was implemented cannot be properly interpreted.

**Comment by kusumotonorio** [comment]
The reason Windows Factor doesn't handle one-codepoint emojis well for display and caret movement is probably because of a data transfer problem with Uniscribe.

For example, encoding  U+1F600  with UTF -16 results in a 4-byte surrogate pair. Uniscribe counts characters as double-byte wide characters and treats them as two characters, but for Windows Factor it is a code point.

I write the quick-hack code below. I cannot display Emoji on Factor in my current environment, so I can't confirm if the problem is solved completely, but I will try it on Windows 10 at a later date.

```
:: line-offset>x ( n script-string -- x )
    n script-string
    2dup string>> length = [
        ssa>> ! ssa
        ! swap 1 - ! icp
        swap 0 swap script-string string>> subseq
        utf16n encode length 2 / >integer 1 - ! icp
        TRUE ! fTrailing
    ] [
        ssa>>
        ! swap ! icp
        swap 0 swap script-string string>> subseq
        utf16n encode length 2 / >integer ! icp
        FALSE ! fTrailing
    ] if
    { int } [ ScriptStringCPtoX check-ole32-error ] with-out-parameters ;
```

```
:: x>line-offset ( x script-string -- n trailing )
    x script-string
    ssa>> ! ssa
    swap ! iX
    { int int } [ ScriptStringXtoCP check-ole32-error ] with-out-parameters
    swap 2 * 0 swap script-string string>> utf16n encode subseq
    utf16n decode length
    swap ;
```

```
: make-ssa ( dc script-string -- ssa )
    dup selection? [ string>> ] when
    !    [ utf16n encode ] ! pString
    !    [ length ] bi ! cString
    utf16n encode ! pString
    dup length 2 / >integer ! cString
    dup 1.5 * 16 + >integer ! cGlyphs -- MSDN says this is "recommended size"
    -1 ! iCharset -- Unicode
    ssa-dwFlags
    0 ! iReqWidth
    f ! psControl
    f ! psState
    f ! piDx
    f ! pTabdef
    f ! pbInClass
    f void* <ref> ! pssa
    [ ScriptStringAnalyse ] keep
    [ check-ole32-error ] [ |ScriptStringFree void* deref ] bi* ;
```

**Comment by kusumotonorio** [comment]
The above code worked as expected on Windows 10 when I pasted multiple "U+1F 600" s.
However, when I typed it using the IME pad, it didn't work properly.

This is probably because when pasted and typed by the user, these characters are different representations in the Factor.

![u+10400](https://user-images.githubusercontent.com/47816570/59957469-97f5f600-94d3-11e9-9c47-51b9460fcfc1.png)

**Comment by kusumotonorio** [comment]
I made the same changes for the Mac Factor that I did for the Windows Factor.
This code seems to resolve the issue https://github.com/factor/factor/issues/1861.

```
USING: io.encodings.utf16n io.encodings.string locals ;

M:: core-text-renderer x>offset ( x font string -- n )
    x font string
    [ 2drop 0 ] [
        cached-line line>>
        swap scale 0 <CGPoint> CTLineGetStringIndexForPosition
    ] if-empty
    2 * 0 swap string utf16n encode subseq
    utf16n decode length ;

M:: core-text-renderer offset>x ( n font string -- x )
    font string cached-line line>> 
    0 n string subseq utf16n encode length 2 / >integer
    f
    CTLineGetOffsetForStringIndex unscale ;
```

![emoji-and-caret](https://user-images.githubusercontent.com/47816570/59962845-c6042600-9525-11e9-80df-1ce60238c728.png)

**Comment by erg** [comment]
I made a branch with the macOS patch and the Windows patch as well as a port of my other branch with unicode 12.1.0 data files.

I'd love to get the branch ``unicode-12.1.0`` working and merged into master.

- [ ] fix collation tests
- [ ] fix emojis on macOS (done?)
- [ ] fix emojis on Windows

**Comment by AlexIljin** [comment]
This is work towards supporting Unicode v12.
How many more Unicode versions, do you think, are to come?

**Comment by erg** [comment]
Many! But we're on ... 5.0 ... so the changes will be hardest this time unless they upend everything someday.

What's the alternative, using a library or just the OS? I haven't researched enough, any ideas?

**Comment by AlexIljin** [comment]
This was not meant as a criticism of our implementation, but rather a poke at the arrogance of the early Unicode efforts, which many deemed to be nearly complete by version 2.

The most complete, mature and free/open-source library is [ICU](https://en.wikipedia.org/wiki/International_Components_for_Unicode), but it's also quite large. I don't like to have huge dependencies, which I don't use or use only a minuscule portion of.

For example, the [latest release](https://github.com/unicode-org/icu/releases/tag/release-64-2) contains a DLL over 26 Mb in size. I don't know if it's a necessary component, or something optional, but I wouldn't want to have to distribute it along with my 6 Mb Factor executables either way.

**Comment by AlexIljin** [comment]
Heck, 6 Mb for what I do is way too much already.

**Comment by kusumotonorio** [comment]
I want the Factor to recognize country flags represented by multiple code points as grapheme clusters.
The Factor codepoint join and separate rules appear to be defined by "make-grapheme-table". 
It seems to me that If we can add new rules appropriately, Factor can make the movement follow unicode rules.

**Comment by kusumotonorio** [comment]
I am testing it. ( https://kusumotonorio.github.io/images/emoji_break.mov )
Emoji Combining Sequence:  U+0031 U+FE0F U+20E3
Emoji Modifier Sequence: U+1F44F U+1F3FD 
Emoji Flag Sequence: U+1F1FA U+1F1F3, U+1F1EF U+1F1F5, U+1F1EA U+1F1FA
Emoji ZWJ Sequence: U+1F468 U+200D U+1F468 U+200D U+1F466, U+1F6B5 U+1F3FC U+200D U+2640 U+FE0F

![emoji_break](https://user-images.githubusercontent.com/47816570/60186103-1fda5800-9866-11e9-84a5-b3b31a45a870.png)

**Comment by kusumotonorio** [comment]
I modified unicode.factor and breaks.factor in order to make Factor work in accordance with the latest [UAX No.29](https://www.unicode.org/reports/tr29/tr29-35.html). I hope it can handle characters that include almost any emoji properly.

What should I do so that you can try it? I can write the contents of those files here, but it's a bit long.

![emoji-england](https://user-images.githubusercontent.com/47816570/60674810-e9c85400-9eb5-11e9-897d-4196585dc11e.png)

**Comment by erg** [comment]
Cool! Can you make a pull request?

**Comment by kusumotonorio** [comment]
Unfortunately, I don't understand much about git or github. How do you do PR for a specific branch?

First, I need to clone branch unicode -12.1. 0 locally, right?

**Comment by erg** [comment]
The usual workflow:

1. fork factor/factor to your github user's account
2. add your fork as a remote to your git directory, ``git remote add kusumotonorio git@github.com:kusumotonorio/factor``
3. check out the unicode-12.1.0 branch locally, commit, and push the changes to your fork  ``git push kusumotonorio HEAD:unicode-12.1.0``
4. visit http://github.com/factor/factor and it should prompt to make a pull request, or just make one by following the github ui

I wish I could relax the permissions on particular forks so you could just push to this fork without approval...

**Comment by mrjbq7** [comment]
If you have a set of patches that work, you can also link to those and we can help you with merging.

**Comment by kusumotonorio** [comment]
Thank you, @erg and @mrjbq7. 
I will try to do PR to allow you to try it this weekend.

By the way, I am sorry that I sometimes write strange sentences. Mistakes I often make are to post English sentences with the wrong subject by automatically converting Japanese sentences without the subject.
Please let me know if you find my writing strange. I'll do it over and over again.

**Comment by kusumotonorio** [comment]
I tried PR. How about that? Please let me know if there is a problem with it.

**Comment by kusumotonorio** [comment]
I did the PR, Inline input using input methods in Mac Factor, for unicode-12.1.0. 
https://github.com/factor/factor/pull/2156

**Comment by erg** [comment]
There are only 12 failing unit tests for ``"unicode.collation"``.

I think they are all problems with the ``ducet`` table.

There are supposed to be 10 extra contractions in the table, but I only added two so far. The way we used to do it caused a test or two to fail. If someone wants a project, try to fix these tests!

```
{ +lt+ } [ { 4019 98 } { 4019 3953 1 3968 97 } [ >string ] bi@ string<=> ] unit-test
{ +lt+ } [ { 4018 820 3969 } { 3959 33 } [ >string ] bi@ string<=> ] unit-test
{ +lt+ } [ { 4019 3953 820 3968 } { 3961 33 } [ >string ] bi@ string<=> ] unit-test


{ { 12748 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 3958 3953 820 } >string collation-key/nfd drop ] unit-test

{ { 12748 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 4018 820 3953 3968 } >string collation-key/nfd drop ] unit-test

{ { 12748 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 4018 820 3968 3953 } >string collation-key/nfd drop ] unit-test

{ { 12748 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 4018 820 3969 } >string collation-key/nfd drop ] unit-test

{ { 12750 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 3960 3953 820 } >string collation-key/nfd drop ] unit-test

{ { 12750 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 4019 820 3953 3968 } >string collation-key/nfd drop ] unit-test

{ { 12750 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 4019 820 3968 3953 } >string collation-key/nfd drop ] unit-test

{ { 12750 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 4019 3953 820 3968 } >string collation-key/nfd drop ] unit-test

{ { 12722 12741 12744 7817 0 32 32 32 32 0 2 2 2 2 0 65535 65535 65535 65535 } }
[ { 4019 3953 1 3968 97 } >string collation-key/nfd drop ] unit-test
```

The ``weight-levels`` below are all wrong, the keys should be correct, according to:
    https://www.unicode.org/reports/tr10/tr10-41.html#Well_Formed_DUCET

We might need a few more key/value pairs as well.


```
! IN: unicode.collation

: fixup-ducet ( -- )
    {
        {
            { 0x0FB2 0x0F71 } ! CE(0FB2) CE(0F71)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F71 } ! CE(0FB3) CE(0F71)
            {
                T{ weight-levels
                    { primary 12720 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }

                ! FIXME: WRONG WEIGHTS
        {
            { 0x0FB2 0x0F71 0x0F72 } ! CE(0FB2) CE(0F71 0F72)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB2 0x0F73        } ! CE(0FB2) CE(0F71 0F72)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB2 0x0F71 0x0F74 } ! CE(0FB2) CE(0F71 0F74)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB2 0x0F75        } ! CE(0FB2) CE(0F71 0F74)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F71 0x0F72 } ! CE(0FB3) CE(0F71 0F72)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F73        } ! CE(0FB3) CE(0F71 0F72)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F71 0x0F74 } ! CE(0FB3) CE(0F71 0F74)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F75        } ! CE(0FB3) CE(0F71 0F74)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
    } ducet get-global '[ swap >string _ set-at ] assoc-each ;
```

**Comment by erg** [comment]
https://collab.its.virginia.edu/wiki/tibetan-script/Sorting%20Tibetan.html

**Comment by erg** [comment]
``{ 3953 3958 3960 3968 3969 } [ canonical-map delete-at ] each``

This makes more of the tests pass, but it's still partially wrong.

**Comment by erg** [comment]
There are basically just three failing collation tests.

The first four  tests normalize (``nfd``) to hex ``{ "fb2" "334" "f71" "f80" }`` or ``{ 4018 820 3953 3968 }``

The second four tests normalize to hex ``{ "fb3" "334" "f71" "f80" }`` or ``{ 4019 820 3953 3968 }``

```
! Second set of failing tests is the same as this test
{ { 12750 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 } }
[ { 3960 3953 820 } >string collation-key/nfd drop ] unit-test


{ 3960 3953 820 } >string nfd >array .
{ 4019 820 3953 3968 } ! nfd
! { "fb3" "334" "f71" "f80" } ! hex nfd

! string>graphemes broken?

! { { 4019 3968 } { 820 } {  3953 }  }
{ { 0xfb3 0xf80 } { 0x334 } { 0xf71 } }
[ >string ] map
graphemes>weights weights>bytes .

! correct answer
{ 12750 12741 0 32 74 32 0 2 2 2 0 65535 65535 65535 }


```

**Comment by erg** [comment]
Commenting out the first two "fixups" in fixup-ducet fixes these tests but breaks a bunch of the other ones.

```
: fixup-ducet ( -- )
    {
        ! breaks first test
        {
            { 0x0FB2 0x0F71 } ! CE(0FB2) CE(0F71)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F71 } ! CE(0FB3) CE(0F71)
            {
                T{ weight-levels
                    { primary 12722 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        ...
```

**Comment by erg** [comment]
```
: fixup-ducet ( -- )
    {
        {
            { 0x0FB2 0x0F71 } ! CE(0FB2) CE(0F71)
            {
                T{ weight-levels
                    { primary 12719 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
        {
            { 0x0FB3 0x0F71 } ! CE(0FB3) CE(0F71)
            {
                T{ weight-levels
                    { primary 12722 }
                    { secondary 32 }
                    { tertiary 2 }
                }
                T{ weight-levels
                    { primary 12741 }
                    { secondary 32 }
                    { tertiary 2 }
                }
            }
        }
```

**Comment by erg** [comment]
The only real issue is if you apply the DUCET fixes for Tibetan then it breaks their tests. Good thing it took me forever to figure this out.

https://www.unicode.org/reports/tr10/tr10-41.html#Well_Formed_DUCET

**Comment by erg** [comment]
I took over this issue for some reason, now I'm questioning if upgrading the unicode files fixes this issue at all. Anyway, we are on 12.1.0 for Unicode data files and collation at least.

Please document what's still broken and open new issues if this one is closed.

**Comment by kusumotonorio** [comment]
I still don't understand "collation" enough to contribute, but I'm learning little by little.

Is the return value of collation-key/nfd appropriate for emoji characters? It is quite different from kanji characters that also have implicit weights.

```
IN: scratchpad "漢" [ first [ AAAA ] [ BBBB ] bi ] [ collation-key/nfd ] bi

--- Data stack:
T{ weight-levels f 64320 32 2 f }
T{ weight-levels f 61218 0 0 f }
{ 64320 61218 0 32 0 2 0 65535 }
"漢"
IN: scratchpad 4drop
IN: scratchpad "😀" [ first [ AAAA ] [ BBBB ] bi ] [ collation-key/nfd ] bi

--- Data stack:
T{ weight-levels f 64451 32 2 f }
T{ weight-levels f 62976 0 0 f }
{ 0 0 0 6030 }
"😀"
```

**Comment by erg** [comment]
```
! This doesn't display well on windows, the face covers up the closing double-quote
{ "漢" "😀" }   natural-sort
```

I hope the collation is correct... basically the collation algorithm lets you sort two strings. There are four levels in weight-levels which are separated by zeros. If two collations have the same first level, they compare the second level, etc. It mostly makes sense within a language, like one of the phonetic Japanese alphabets, but not so much when comparing an emoji with language. (I'm sure it's well defined by the collation algorithm, it just doesn't mean much to me if an emoji should come before or after a word)

**Comment by kusumotonorio** [comment]
I was also concerned that Emoji does not display in color on Windows. If we use Uniscribe, GDI, we can't display the emoji in color. I saw the information that we need to use Direct2D to achieve color display.

**Comment by kusumotonorio** [comment]
Operations that work well with Windows7 do not seem to work well with Windows10. Sometimes it results in a Factor outage.
My windows7 is 32bit and windows10 is 64bit, so I can not replace usp10.dll and test it. I will try to create a more appropriate patch for Windows 10, but your attempts are also welcome.


![cntrl_a](https://user-images.githubusercontent.com/47816570/62177838-677e6300-b381-11e9-8cb2-14f55b189fa6.png)

**Comment by erg** [comment]
Windows bug above is fixed.

Maybe we can close this now by saying length is code-point length and if you want grapheme length just do ``>graphemes length``

```factor
"🇪🇸" length .
2

"🇪🇸" >graphemes length .
1
```

**Comment by kusumotonorio** [comment]
Or use the word which I add, ```count-graphemes```. ;-)

The first issue with @mrjbq7 mrjbq7 is that string length should return the number of glyphs, but I think it is appropriate to return the number of code points.
As @erg has appropriately revised the title of this issue, the question is where we stand.
The Factor now follows the latest unicode standard for determining the number of graphemes, but the number of glyphs displayed depends on the text rendering engine used by the Factor for each platform. 

I don't think the summary displayed when I inspect a string is very good.

For example, ```Summary: string with 7 characters```. 
https://github.com/factor/factor/issues/1402#issuecomment-508509008

If this "characters" is a glyph, this is not always true. However, if this appears as ```Summary: string with 7 code points```, then it is correct.

**Comment by mrjbq7** [comment]
I'd be in favor of changing it to code points, that is what it is, good idea!

**Comment by erg** [comment]
@mrjbq7  Can you change the summary string and add a ``count-graphemes`` word and we can close this issue?

**Comment by mrjbq7** [comment]
Fixed in b0390c57aeff276683589b68ed3574156420778a.

The ``count-graphemes`` word already exists (and is currently implemented as``>graphemes length``).
